packages/gestalt/src/MasonryV2.js

```diff
@@ -0,0 +1,654 @@
+  scrollTop,
+  virtualBufferFactor,
+  virtualBoundsTop,
+  virtualBoundsBottom,
+  virtualize,
+}: {
+  containerHeight: number,
+  containerOffset: number,
+  scrollContainer: ?HTMLElement,
+  scrollTop: number,
+  virtualBufferFactor: number,
+  virtualBoundsTop: ?number,
+  virtualBoundsBottom: ?number,
+  virtualize: boolean,
+}) {
+  if (virtualize && scrollContainer && Boolean(virtualBufferFactor)) {
+    const virtualBuffer = containerHeight * virtualBufferFactor;
+    const offsetScrollPos = scrollTop - containerOffset;
+    const viewportTop = virtualBoundsTop
+      ? offsetScrollPos - virtualBoundsTop
+      : offsetScrollPos - virtualBuffer;
+    const viewportBottom = virtualBoundsBottom
+      ? offsetScrollPos + containerHeight + virtualBoundsBottom
+      : offsetScrollPos + containerHeight + virtualBuffer;
+
+    return {
+      viewportTop,
+      viewportBottom,
+    };
+  }
+  return {
+    viewportTop: -Infinity,
+    viewportBottom: Infinity,
+  };
+}
+
+function Masonry<T: { +[string]: mixed }>(
+  {
+    columnWidth = 236,
+    gutterWidth: gutter,
+    items,
+    layout = 'basic',
+    loadItems = () => {},
+    measurementStore: measurementStoreProp,
+    minCols = 3,
+    positionStore: positionStoreProp,
+    renderItem,
+    scrollContainer,
+    virtualBufferFactor = 0.7,
+    virtualBoundsBottom,
+    virtualBoundsTop,
+    virtualize = false,
+    _twoColItems,
+    _logTwoColWhitespace,
+    _measureAll,
+  }: Props<T>,
+  ref: { current: null | MasonryRef, ... } | ((null | MasonryRef) => mixed),
+): ReactNode {
+  const hasSetInitialWidth = useRef(false);
+  const [gridWrapperEl, setGridWrapperEl] = useState<?HTMLDivElement>(null);
+  const gridWrapperRef = useCallback((el: ?HTMLDivElement) => {
+    if (el) {
+      setGridWrapperEl(el);
+    }
+  }, []);
+
+  const heightsStore = useMemo(() => new HeightsStore(), []);
+
+  const measurementStore = useMemo(
+    () => measurementStoreProp || createMeasurementStore(),
+    [measurementStoreProp],
+  );
+
+  const positionStore = useMemo(
+    () => positionStoreProp || createMeasurementStore(),
+    [positionStoreProp],
+  );
+
+  const scrollContainerElement = useMemo(
+    () => (typeof window !== 'undefined' ? scrollContainer?.() : null),
+    [scrollContainer],
+  );
+
+  const width = useElementWidth(gridWrapperEl);
+  const { containerHeight, containerOffset, scrollTop } = useScrollContainer({
+    gridWrapper: gridWrapperEl,
+    scrollContainer: scrollContainerElement,
+  });
+
+  const [, startTransition] = useTransition();
+  const forceUpdate = useForceUpdate();
+
+  const reflow = () => {
+    measurementStore.reset();
+    positionStore.reset();
+    heightsStore.reset();
+
+    startTransition(() => {
+      forceUpdate();
+    });
+  };
+
+  // these are all for backwards compatibility with the old Masonry
+  // will work on removing these once this lands
+  useImperativeHandle(ref, () => ({
+    handleResize: () => reflow(),
```

Comment:
there are some use cases in pinboard which currently reference the Masonry class instance - these were the ones I found, so currently exposing on the ref for backwards compatibility. will lock down what's exposed on the ref on the Pinboard side and also remove these whenever this ships